ktor-network/posix/src/io/ktor/network/sockets/BaseNativeSocket.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.util.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.errors.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import platform.posix.*
+import kotlin.coroutines.*
+import kotlin.math.*
+
+internal abstract class BaseNativeSocket(
```

Comment:
I would try extracting common parts in the separate utility methods instead of creating abstract `Base` here.

Reply:
Done. Please check out the new `CIOReader`/`CIOWriter` files. The `BaseNativeSocket` class is now removed. I tried to keep the naming similar to the JVM datagram sockets implementation.